File: src/main/java/com/github/sorusclient/client/hud/impl/health/v1_8_9/HealthRenderer.java
package com.github.sorusclient.client.hud.impl.health.v1_8_9;

import com.github.glassmc.loader.GlassLoader;
import com.github.glassmc.loader.Listener;
import com.github.sorusclient.client.hud.impl.health.IHealthRenderer;
import org.lwjgl.opengl.GL11;
import v1_8_9.net.minecraft.client.MinecraftClient;
import v1_8_9.net.minecraft.client.gui.DrawableHelper;
import v1_8_9.net.minecraft.util.Identifier;

public class HealthRenderer implements Listener, IHealthRenderer {

    @Override
    public void run() {
        GlassLoader.getInstance().registerInterface(IHealthRenderer.class, this);
    }

    @Override
    public void renderHeart(double x, double y, double scale, HeartType heartType, HeartRenderType heartRenderType) {
        GL11.glPushMatrix();
        GL11.glTranslated(x, y, 0);
        GL11.glScaled(scale, scale, 1);

        DrawableHelper drawableHelper = new DrawableHelper();
        GL11.glColor4d(1.0, 1.0, 1.0, 1.0);

        MinecraftClient.getInstance().getTextureManager().bindTexture(new Identifier("textures/gui/icons.png"));

        int xLocation = heartType == HeartType.HEALTH ? 52 : 52 + 12 * 9;

        if (heartRenderType == HeartRenderType.FULL) {
            drawableHelper.drawTexture(0, 0, xLocation, 0, 9, 9);
        } else if (heartRenderType == HeartRenderType.HALF_EMPTY) {
            drawableHelper.drawTexture(0, 0, xLocation + 9, 0, 9, 9);
        } else if (heartRenderType == HeartRenderType.HALF_DAMAGE) {
            drawableHelper.drawTexture(0, 0, xLocation + 18, 0, 9, 9);
            drawableHelper.drawTexture(0, 0, xLocation + 9, 0, 9, 9);
        } else if (heartRenderType == HeartRenderType.DAMAGE) {
            drawableHelper.drawTexture(0, 0, xLocation + 18, 0, 9, 9);
        } else if (heartRenderType == HeartRenderType.DAMAGE_EMPTY) {
            drawableHelper.drawTexture(0, 0, xLocation + 27, 0, 9, 9);
        }

        GL11.glPopMatrix();
    }

    @Override
    public void renderHeartBackground(double x, double y, double scale, BackgroundType backgroundType) {
        GL11.glPushMatrix();
        GL11.glTranslated(x, y, 0);
        GL11.glScaled(scale, scale, 1);

        DrawableHelper drawableHelper = new DrawableHelper();
        GL11.glColor4d(1.0, 1.0, 1.0, 1.0);

        MinecraftClient.getInstance().getTextureManager().bindTexture(new Identifier("textures/gui/icons.png"));

        switch (backgroundType) {
            case STANDARD:
                drawableHelper.drawTexture(0, 0, 16, 0, 9, 9);
                break;
            case FLASHING_OUTLINE:
                drawableHelper.drawTexture(0, 0, 25, 0, 9, 9);
                break;
        }
        GL11.glPopMatrix();
    }

}

File: src/main/kotlin/com/github/sorusclient/client/hud/impl/health/v1_8_9/HealthRenderer.kt
package com.github.sorusclient.client.hud.impl.health.v1_8_9

import com.github.glassmc.loader.api.GlassLoader
import com.github.glassmc.loader.api.Listener
import com.github.sorusclient.client.hud.impl.health.IHealthRenderer
import com.github.sorusclient.client.hud.impl.health.IHealthRenderer.BackgroundType
import com.github.sorusclient.client.hud.impl.health.IHealthRenderer.HeartType
import org.lwjgl.opengl.GL11
import v1_8_9.net.minecraft.client.MinecraftClient
import v1_8_9.net.minecraft.client.render.Tessellator
import v1_8_9.net.minecraft.client.render.VertexFormats
import v1_8_9.net.minecraft.util.Identifier

class HealthRenderer : Listener, IHealthRenderer {
    override fun run() {
        GlassLoader.getInstance().registerInterface(IHealthRenderer::class.java, this)
    }

    override fun renderHeart(
        x: Double,
        y: Double,
        scale: Double,
        heartType: HeartType,
        heartRenderType: IHealthRenderer.HeartRenderType
    ) {
        GL11.glPushMatrix()
        GL11.glTranslated(x, y, 0.0)
        GL11.glScaled(scale, scale, 1.0)
        GL11.glColor4d(1.0, 1.0, 1.0, 1.0)
        GL11.glEnable(GL11.GL_BLEND)
        MinecraftClient.getInstance().textureManager.bindTexture(Identifier("textures/gui/icons.png"))
        val xLocation = if (heartType == HeartType.HEALTH) 52 else 52 + 12 * 9
        if (heartRenderType == IHealthRenderer.HeartRenderType.FULL) {
            drawTexture(0, 0, xLocation.toDouble(), 0.0, 9, 9)
        } else if (heartRenderType == IHealthRenderer.HeartRenderType.HALF_EMPTY) {
            drawTexture(0, 0, xLocation.toDouble() + 9, 0.0, 9, 9)
        } else if (heartRenderType == IHealthRenderer.HeartRenderType.HALF_DAMAGE) {
            drawTexture(0, 0, xLocation.toDouble() + 18, 0.0, 9, 9)
            drawTexture(0, 0, xLocation.toDouble() + 9, 0.0, 9, 9)
        } else if (heartRenderType == IHealthRenderer.HeartRenderType.DAMAGE) {
            drawTexture(0, 0, xLocation.toDouble() + 18, 0.0, 9, 9)
        } else if (heartRenderType == IHealthRenderer.HeartRenderType.DAMAGE_EMPTY) {
            drawTexture(0, 0, xLocation.toDouble() + 27, 0.0, 9, 9)
        }
        GL11.glPopMatrix()
    }

    override fun renderHeartBackground(x: Double, y: Double, scale: Double, backgroundType: BackgroundType?) {
        GL11.glPushMatrix()
        GL11.glTranslated(x, y, 0.0)
        GL11.glScaled(scale, scale, 1.0)
        GL11.glColor4d(1.0, 1.0, 1.0, 1.0)
        GL11.glEnable(GL11.GL_BLEND)
        MinecraftClient.getInstance().textureManager.bindTexture(Identifier("textures/gui/icons.png"))
        when (backgroundType) {
            BackgroundType.STANDARD -> drawTexture(0, 0, 16.0, 0.0, 9, 9)
            BackgroundType.FLASHING_OUTLINE -> drawTexture(0, 0, 25.0, 0.0, 9, 9)
        }
        GL11.glPopMatrix()
    }

    private fun drawTexture(var1: Int, var2: Int, var3: Double, var4: Double, var5: Int, var6: Int) {
        val var3 = var3 + 0.1
        val var9 = Tessellator.getInstance()
        val var10 = var9.buffer
        var10.begin(7, VertexFormats.POSITION_TEXTURE)
        var10.vertex((var1 + 0).toDouble(), (var2 + var6).toDouble(), 0.0).texture(((var3 + 0).toFloat() / 256).toDouble(), ((var4 + var6).toFloat() / 256).toDouble()).next()
        var10.vertex((var1 + var5).toDouble(), (var2 + var6).toDouble(), 0.0).texture(((var3 + var5).toFloat() / 256).toDouble(), ((var4 + var6).toFloat() / 256).toDouble()).next()
        var10.vertex((var1 + var5).toDouble(), (var2 + 0).toDouble(), 0.0).texture(((var3 + var5).toFloat() / 256).toDouble(), ((var4 + 0).toFloat() / 256).toDouble()).next()
        var10.vertex((var1 + 0).toDouble(), (var2 + 0).toDouble(), 0.0).texture(((var3 + 0).toFloat() / 256).toDouble(), ((var4 + 0).toFloat() / 256).toDouble()).next()
        var9.draw()
    }

}